common/scala/src/main/scala/org/apache/openwhisk/core/database/elasticsearch/ElasticSearchActivationStore.scala

```diff
@@ -0,0 +1,482 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.openwhisk.core.database.elasticsearch
+
+import java.time.Instant
+
+import scala.language.postfixOps
+import akka.actor.ActorSystem
+import akka.event.Logging.ErrorLevel
+import akka.http.scaladsl.model._
+import akka.stream.scaladsl.Flow
+import akka.stream._
+import com.sksamuel.elastic4s.http.{ElasticClient, ElasticProperties, NoOpRequestConfigCallback}
+import com.sksamuel.elastic4s.indexes.IndexRequest
+import com.sksamuel.elastic4s.searches.queries.matches.MatchPhrase
+import org.apache.http.auth.{AuthScope, UsernamePasswordCredentials}
+import org.apache.http.impl.client.BasicCredentialsProvider
+import org.apache.http.impl.nio.client.HttpAsyncClientBuilder
+import pureconfig.loadConfigOrThrow
+import spray.json._
+import org.apache.openwhisk.common.{Logging, LoggingMarkers, TransactionId}
+import org.apache.openwhisk.core.ConfigKeys
+import org.apache.openwhisk.core.containerpool.logging.ElasticSearchJsonProtocol._
+import org.apache.openwhisk.core.database._
+import org.apache.openwhisk.core.entity._
+import org.apache.openwhisk.http.Messages
+import org.elasticsearch.client.RestClientBuilder.HttpClientConfigCallback
+
+import scala.concurrent.{ExecutionContextExecutor, Future, Promise}
+import scala.concurrent.duration._
+import scala.util.Try
+
+case class ElasticSearchActivationStoreConfig(protocol: String,
+                                              hosts: String,
+                                              indexPattern: String,
+                                              username: String,
+                                              password: String)
+
+class ElasticSearchActivationStore(
+  httpFlow: Option[Flow[(HttpRequest, Promise[HttpResponse]), (Try[HttpResponse], Promise[HttpResponse]), Any]] = None,
+  elasticSearchConfig: ElasticSearchActivationStoreConfig =
+    loadConfigOrThrow[ElasticSearchActivationStoreConfig](ConfigKeys.elasticSearchActivationStore),
+  useBatching: Boolean = false)(implicit actorSystem: ActorSystem,
+                                actorMaterializer: ActorMaterializer,
+                                logging: Logging)
+    extends ActivationStore
+    with MultipleReadersSingleWriterCache[WhiskActivation, DocInfo] {
```

Comment:
Does the same activation record gets fetched multiple times for the cache to be effective?

Reply:
hm, maybe by `wsk activation get $aid`? seems this is randomly used

Reply:
That would be a developer flow and does not come in critical path. In most cases the activation record should be consumed directly (if blocking) or via the first call for poll. Do not think that same activation record would be fetched multiple times. 

So caching it should not be required. Note we currently do not cache it for other existing db stores also